lsp/nls/src/linearization.rs

```diff
@@ -80,6 +128,146 @@ impl BuildingExt for Linearization<Building<BuildingResource>> {
+                .and_then(|parent_declaration| self.resolve_reference(parent_declaration))
+                // get record field
+                .and_then(|parent_declaration| match &parent_declaration.kind {
+                    TermKind::Record(fields) => {
+                        fields.get(&child_ident).and_then(|child_declaration_id| {
+                            self.state.resource.linearization.get(*child_declaration_id)
+                        })
+                    }
+                    _ => None,
+                });
+
+            let referenced_declaration =
+                referenced_declaration.and_then(|referenced| match referenced.kind {
+                    TermKind::Usage(UsageState::Resolved(Some(pointed))) => {
+                        self.state.resource.linearization.get(pointed)
+                    }
+                    TermKind::RecordField { value, .. } => value
+                        // retrieve record
+                        .and_then(|value_index| self.state.resource.linearization.get(value_index))
+                        // retrieve field
+                        .and_then(|record| match &record.kind {
+                            TermKind::Record(fields) => {
+                                debug!(
+                                    "parent referenced a nested record indirectly`: {:?}",
+                                    fields
+                                );
+                                fields.get(&child_ident).and_then(|accessor_id| {
+                                    self.state.resource.linearization.get(*accessor_id)
+                                })
+                            }
+                            TermKind::Usage(UsageState::Resolved(Some(pointed))) => {
+                                self.state.resource.linearization.get(*pointed)
+                            }
+                            _ => None,
+                        })
+                        .or(Some(referenced)),
+
+                    _ => Some(referenced),
+                });
+
+            let referenced_id = referenced_declaration.map(|reference| reference.id);
+
+            debug!(
+                "Associating child {} to value {:?}",
+                child_ident, referenced_declaration
+            );
+
+            {
+                let child: &mut LinearizationItem<TypeWrapper> = self
+                    .state
+                    .resource
+                    .linearization
+                    .get_mut(*child_item)
+                    .unwrap();
+                child.kind = TermKind::Usage(UsageState::Resolved(referenced_id));
+            }
+
+            if let Some(referenced_id) = referenced_id {
+                self.add_usage(referenced_id, *child_item);
+            }
+        }
+
+        if !unresolved.is_empty() {
+            debug!("unresolved references: {:?}", unresolved);
+            self.resolve_record_references(&unresolved);
```

Comment:
It seems the following is happening:

- the caller of `resolve_record_references` is building a vector, that is then passed the function as a slice. But is not used thereafter by the caller.
- this function loops on the slice, and build a second vector, and then calls itself recursively.

To avoid a bit of fuss, it may be simpler to just pass an owned vector, that you populate and pop in a while loop (or a queue, if you need to preserve order while iterating, or a pointer to a vector). This is not really important though, just a suggestion, to avoid recursion, as you already have to loop anyway.

Reply:
Yes that would be a clearer solution.